matcher/src/pattern.rs

```diff
@@ -0,0 +1,476 @@
+                let pat = Atom::parse(pat, case_matching);
+                (!pat.needle.is_empty()).then_some(pat)
+            })
+            .collect();
+        Pattern { atoms }
+    }
+
+    /// Convenience function to easily match (and sort) a (relatively small)
+    /// list of inputs.
+    ///
+    /// *Note* This unction is not recommended for building a full fuzzy
+    /// matching application that can match large numbers of matches (like all
+    /// files in a directory) as all matching is done on the current thread,
+    /// effectively blocking the UI. For such applications the high level
+    /// `nucleo` crate can be used instead.
+    pub fn match_list<T: AsRef<str>>(
+        &self,
+        items: impl IntoIterator<Item = T>,
+        matcher: &mut Matcher,
+    ) -> Vec<(T, u32)> {
+        if self.atoms.is_empty() {
+            return items.into_iter().map(|item| (item, 0)).collect();
+        }
+        let mut buf = Vec::new();
+        let mut items: Vec<_> = items
+            .into_iter()
+            .filter_map(|item| {
+                self.score(Utf32Str::new(item.as_ref(), &mut buf), matcher)
+                    .map(|score| (item, score))
+            })
+            .collect();
+        items.sort_by_key(|(_, score)| Reverse(*score));
+        items
+    }
+
+    /// Matches this pattern against `haystack` (using the allocation and configuration
+    /// from `matcher`) and calculates a ranking score. See the [`Matcher`](crate::Matcher).
+    /// Documentation for more details.
+    ///
+    /// *Note:*  The `ignore_case` setting is overwritten to match the casing of
+    /// each pattern atom.
+    pub fn score(&self, haystack: Utf32Str<'_>, matcher: &mut Matcher) -> Option<u32> {
+        if self.atoms.is_empty() {
+            return Some(0);
+        }
+        let mut score = 0;
+        for pattern in &self.atoms {
+            score += pattern.score(haystack, matcher)? as u32;
+        }
+        Some(score)
+    }
+
+    /// Matches this pattern against `haystack` (using the allocation and
+    /// configuration from `matcher`), calculates a ranking score and the matche
+    /// indices. See the [`Matcher`](crate::Matcher). Documentation for more
+    /// details.
+    ///
+    /// *Note:*  The `ignore_case` setting is overwritten to match the casing of
+    /// each pattern atom.
+    ///
+    /// *Note:*  The indices for each pattern are calculated individually
+    /// and simply appended to the `indices` vector. This allows
+    ///
+    pub fn indices(
+        &self,
+        haystack: Utf32Str<'_>,
+        matcher: &mut Matcher,
+        indices: &mut Vec<u32>,
+    ) -> Option<u32> {
+        if self.atoms.is_empty() {
+            return Some(0);
+        }
+        let mut score = 0;
+        for pattern in &self.atoms {
+            score += pattern.indices(haystack, matcher, indices)? as u32;
+        }
+        Some(score)
+    }
+
+    /// Refreshes this pattern by reparsing a
```

Comment:
Same here

Reply:
Ah i think I accidently closed helix without saving once (by closing the emultaor) so that explains the cutoff doc comments :D